types/time.go

```diff
@@ -1338,6 +1348,41 @@ func checkDateType(t TimeInternal, allowZeroInDate bool) error {
 	return nil
 }
 
+func checkDateTypeInModify(t TimeInternal, sc *variable.StatementContext) error {
+	year, month, day := t.Year(), t.Month(), t.Day()
+	strict, noZeroDate, noZeroInDate, ignoreErr := true, true, true, false
+	if sc != nil && (sc.InInsertStmt || sc.InUpdateOrDeleteStmt) {
+		m := sc.SQLMode
+		strict, noZeroDate, noZeroInDate, ignoreErr = m.HasStrictMode(), m.HasNoZeroDateMode(), m.HasNoZeroInDateMode(), sc.IgnoreErr
+	}
+	if year == 0 && month == 0 && day == 0 {
+		if !noZeroDate {
```

Comment:
Double negative make the logic hard to read.

Reply:
ok. I'll try to refine it.